atomisticparsers/h5md/parser.py

```diff
@@ -0,0 +1,787 @@
+                sec_calc = sec_run.m_create(Calculation)
+                sec_calc.time = frame * time_unit
+            for calc_entry in data_h5md['x_h5md_custom_calculations']:
+                sec_calc.x_h5md_custom_calculations.append(calc_entry)
+            sec_energy = sec_calc.energy
+            if not sec_energy:
+                sec_energy = sec_calc.m_create(Energy)
+            for energy_entry in data_h5md['x_h5md_energy_contributions']:
+                sec_energy.x_h5md_energy_contributions.append(energy_entry)
+
+    def parse_system(self):
+        sec_run = self.archive.run[-1]
+
+        system_info = self.system_info.get('system')
+        if not system_info:
+            self.logger.error('No particle information found in H5MD file.')
+
+        n_frames = len(system_info.get('times', []))
+        self._system_time_map = {}
+        self._system_step_map = {}
+        for frame in range(n_frames):
+            # if (n % self.frame_rate) > 0:
+            #     continue
+
+            atoms_dict = {}
+            for key in system_info.keys():
+                if key == 'times':
+                    time = system_info.get('times', [None] * (frame + 1))[frame]
+                    if time is not None:
+                        self._system_time_map[round(ureg.convert(
+                            time.magnitude, time.units, ureg.picosecond), 5)] = len(self._system_time_map)
+                elif key == 'steps':
+                    step = system_info.get('steps', [None] * (frame + 1))[frame]
+                    if step is not None:
+                        self._system_step_map[round(step)] = len(self._system_step_map)
+                else:
+                    atoms_dict[key] = system_info.get(key, [None] * (frame + 1))[frame]
+
+            if frame == 0:  # TODO extend to time-dependent bond lists
+                connectivity = self.h5md_groups['connectivity']
+                atoms_dict['bond_list'] = self._data_parser.hdf5_getter(connectivity, 'bonds')
+
+            self.parse_trajectory_step({
+                'atoms': atoms_dict
+            })
+
+            if frame == 0:  # TODO extend to time-dependent topologies
+                topology = self._data_parser.hdf5_getter(connectivity, 'particles_group', None)
+                if topology:
+                    self.get_atomsgroup_fromh5md(sec_run.system[frame], topology)
+
+    def parse_method(self):
+
+        sec_method = self.archive.run[-1].m_create(Method)
+        sec_force_field = sec_method.m_create(ForceField)
+        sec_model = sec_force_field.m_create(Model)
+
+        # get the atom parameters
+        n_atoms = self.n_atoms[0]  # TODO Extend to non-static n_atoms
+        for n in range(n_atoms):
+            sec_atom = sec_method.m_create(AtomParameters)
+            for key in self.atom_parameters.keys():
+                setattr(sec_atom, key, self.atom_parameters[key][n])
```

Comment:
do not use setattr, use instead sec_atom.m_set(sec_atom.get_quantity_attribute(key), self.atom_parameters[key][n])

Reply:
Should I also do instead of m_create?:
            sec_atom = AtomParameters()
            sec_method.atom_parameters.append(sec_atom)

Reply:
yes

Reply:
Also, with m_set, I get the error:
```
self = AtomParameters(), name = 'get_quantity_attribute'

    def __getattr__(self, name):
        # The existence of __getattr__ will make mypy and pylint ignore 'missing' dynamic
        # attributes and functions and wrong types of those.
        # Ideally we have a plugin for both that add the correct type info
    
        if name == 'm_key':
            return self.get(name)
    
        if name in self.m_def.all_aliases:
            return getattr(self, self.m_def.all_aliases[name].name)
    
        if self.m_def.has_variable_names:
            m_definition: Definition = resolve_variadic_name(self.m_def.all_properties, name)
            if m_definition:
                if not isinstance(m_definition, Quantity) or not m_definition.use_full_storage:
                    return getattr(self, m_definition.name)
    
                m_storage: dict = self.__dict__.get(m_definition.name, None)
                if m_storage is None:
                    return None
    
                m_quantity = m_storage.get(name, None)
                if m_quantity is None:
                    return None
    
                if m_quantity.value is not None:
                    if m_quantity.unit is not None:
                        return units.Quantity(m_quantity.value, m_quantity.unit)
    
                    return m_quantity.value
    
        if name.startswith('a_'):
            annotation_name = name[2:]
            if annotation_name not in self.m_annotations:
                raise AttributeError(name)
    
            return self.m_get_annotations(annotation_name)
    
>       raise AttributeError(name)
E       AttributeError: get_quantity_attribute

../../../nomad/metainfo/metainfo.py:1308: AttributeError
```

It's failing for `key = "label"`, which is a quantity in AtomParameters(). Any idea what is wrong here?

Reply:
sorry, it should be m_get_quantity_defintion()

Reply:
ah, yes, that's right, thanks. I thought I had used this before, but I couldn't find an example. done.